charmcraft/providers/_multipass.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Build environment provider support for charmcraft."""
+
+import contextlib
+import logging
+import pathlib
+import re
+from typing import List
+
+from craft_providers import multipass
+
+from charmcraft.cmdbase import CommandError
+from charmcraft.config import Base
+from charmcraft.env import get_managed_environment_project_path
+from charmcraft.utils import confirm_with_user, get_host_architecture
+
+from ._buildd import BASE_CHANNEL_TO_BUILDD_IMAGE_ALIAS, CharmcraftBuilddBaseConfiguration
+from ._provider import Provider
+
+logger = logging.getLogger(__name__)
+
+
+class MultipassProvider(Provider):
+    """Charmcraft's build environment provider.
+
+    :param multipass: Optional Multipass client to use.
+    """
+
+    def __init__(
+        self,
+        *,
+        multipass: multipass.Multipass = multipass.Multipass(),
+    ) -> None:
+        self.multipass = multipass
+
+    def clean_project_environments(
+        self,
+        *,
+        charm_name: str,
+        project_path: pathlib.Path,
+    ) -> List[str]:
+        """Clean up any environments created for project.
+
+        :param charm_name: Name of project.
+        :param project_path: Directory of charm project.
+
+        :returns: List of containers deleted.
+        """
+        deleted: List[str] = []
+
+        # Nothing to do if provider is not installed.
+        if not self.is_provider_available():
+            return deleted
+
+        inode = str(project_path.stat().st_ino)
+
+        for name in self.multipass.list():
+            match_regex = f"^charmcraft-{charm_name}-{inode}-.+-.+-.+$"
+            if re.match(match_regex, name):
+                logger.debug("Deleting Multipass VM %r.", name)
+                self.multipass.delete(
```

Comment:
This is always successful? May we hit the situation of deleting one env, failing to delete the second, and need to delete the third?

Reply:
this should be try, and catched with a message

Reply:
will fix

Reply:
Fixed for both this and list. (and added tests)

Reply:
Fixed!